### PR TITLE
Add TensorRT optimization profiles

### PR DESCRIPTION
**Pull request by tfeher**
This PR adds optimization profile handling to TF-TRT and enables inference using the experimental dynamic shapes mode (where dynamic shapes mode =  explicit batch mode with dynamic input shapes).

This PR is not supposed to change the existing behavior in implicit batch mode. 

For each unknown input dimension, a shape [optimization profile](https://docs.nvidia.com/deeplearning/sdk/tensorrt-api/c_api/classnvinfer1_1_1_i_optimization_profile.html) describes what is the expected [min, max] range for that dimension, and also define an optimum value. "The builder selects the kernels that result in the lowest runtime for the optimum input tensor dimensions, and are valid for all input tensor sizes in the valid range between minimum and maximum dimensions."

To create an engine with dynamic shapes one must define at least one optimization profile. In the current PR we define only a single optimization profile, using the shape of the concrete input data during the [first inference call](https://github.com/tfeher/tensorflow/blob/1ab228164fcbc648a8b885e2d05ccbf68375758f/tensorflow/compiler/tf2tensorrt/kernels/trt_engine_op.cc#L599-L607).

The real advantage of dynamic shapes is to have a single engine that can execute the network with different input shapes. Unfortunately it is not possible to demonstrate this with the current PR. Although the `TrtShapeOptimizationProfile` class is designed to manage multiple profiles, we still need to enable profile generation mode during build phase, to take the full advatage of dynamic shapes mode. That feature is implemented in PR #36080.

### CODE REVIEW

**Comment by gbaned** [comment]
@tfeher Can you please resolve conflicts? Thanks!

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc [hunk 1]
```diff
@@ -249,6 +250,19 @@ void GetInputProperties(const grappler::GraphProperties& graph_properties,
   }
 }
 
+// This function checks if a tensor is compatible with TRT.
+//
+// We check that the shape and datatype is compatible with TensorRT. We also
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc]: Replace "is" with "are"?

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc [hunk 2]
```diff
@@ -249,6 +250,19 @@ void GetInputProperties(const grappler::GraphProperties& graph_properties,
...
+// return the corresponding trt_dtype, the trt_dims and the batch_size (latter
+// is only needed in implicit batch mode).
+//
+// The return status indicates wether the tensor is compatible.
+//
+// If validation_only == false, then we make an additional check. In implicit
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc]: Can we simplify this paragraph?
For implicit batch mode, when validation_only == false, we also check that all input dimensions besides the implicit batch dimension, are known dimensions.

File: tensorflow/compiler/tf2tensorrt/convert/utils.h [hunk 3]
```diff
@@ -106,6 +106,11 @@ string GetLinkedTensorRTVersion();
 // TensorRT library version information {Maj, Min, Patch}.
 string GetLoadedTensorRTVersion();
 
+// Returns the number of inputs for the engine, which also correspends to the
+// number of input tensors for the network. This can differ from the number of
+// input bindings, because each profile has a set of bindings.
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/convert/utils.h]: Shall we replace the second sentence here with this:
The number of total input bindings equals the number of profile times the number of engine inputs.

File: tensorflow/compiler/tf2tensorrt/kernels/trt_engine_op.cc [hunk 4]
```diff
@@ -92,7 +93,7 @@ class TRTEngineOp : public AsyncOpKernel {
       LRUCache<std::vector<TensorShape>, std::unique_ptr<EngineContext>,
                VectorTensorShapeHasher>;
 
-  // Execute calibration
+  // Execute calibration.
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/kernels/trt_engine_op.cc]: Executes

File: tensorflow/compiler/tf2tensorrt/kernels/trt_engine_op.cc [hunk 5]
```diff
@@ -108,9 +109,10 @@ class TRTEngineOp : public AsyncOpKernel {
 
   // Execute the tensorrt engine. Returns whether we need to retry by running
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/kernels/trt_engine_op.cc]: Executes

File: tensorflow/compiler/tf2tensorrt/kernels/trt_engine_op.cc [hunk 6]
```diff
@@ -108,9 +109,10 @@ class TRTEngineOp : public AsyncOpKernel {
...
   // the native segment.
-  bool ExecuteTrtEngine(OpKernelContext* ctx, EngineContext* engine_context);
+  bool ExecuteTrtEngine(OpKernelContext* ctx, EngineContext* engine_context,
+                        int trt_context_idx);
 
-  // Allocate necessary resources for calibration
+  // Allocate necessary resources for calibration.
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/kernels/trt_engine_op.cc]: Allocates

File: tensorflow/compiler/tf2tensorrt/kernels/trt_engine_op.cc [hunk 7]
```diff
@@ -594,19 +596,33 @@ void TRTEngineOp::ComputeAsync(OpKernelContext* ctx,
 
   OP_REQUIRES_OK_ASYNC(ctx, VerifyInputShapes(input_concrete_shapes), *helper);
 
+  if (!use_implicit_batch_) {
+    if (cache_res->profiles_.GetNumProfiles() == 0) {
+      // Create a single profile from the current input shape.
+      // In the future we will collect a set of input shapes during build mode
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/kernels/trt_engine_op.cc]: Please avoid partial lines in comments.

File: tensorflow/compiler/tf2tensorrt/kernels/trt_engine_op_test.cc [hunk 8]
```diff
@@ -225,16 +225,7 @@ TEST_F(TRTEngineOpTestBase, DynamicShapes) {
...
-  // the engine with optimization profiles to use dynamic input shapes, but that
-  // feature is not yet implemented.
-  //
-  // Since TRT engine creation has failed, we fall back to native segment.
-  // Calling the native segment fails for the same reason that is investigated
-  // in https://github.com/tensorflow/tensorflow/pull/34919. This is irrelevant
-  // for the current test, here we want to just check wether TRT engine creation
-  // has failed.
-  OpsTestBase::RunOpKernel();
+  TF_ASSERT_OK(OpsTestBase::RunOpKernel());
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/kernels/trt_engine_op_test.cc]: Can you add #ifdef and make sure that the test "pass" with all TRT version?

File: tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles.cc [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+
+#include "tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles.h"
+#include <algorithm>
+#include <functional>
+#include "tensorflow/compiler/tf2tensorrt/convert/utils.h"
+
+namespace tensorflow {
+namespace tensorrt {
+
+// Create optimization profiles for a list of input shapes. The list of input
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles.cc]: Creates

File: tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles_test.cc [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+        builder_->createBuilderConfig());
+    builder_config_->setMaxWorkspaceSize(1 << 10);
+#else
+    network_ = TrtUniquePtrType<nvinfer1::INetworkDefinition>(
+        builder_->createNetwork());
+    builder_->setMaxWorkspaceSize(1 << 10);
+#endif
+  }
+
+  // Define a simple network: output = input1 + input2.
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles_test.cc]: Defines

File: tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles.h [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+      std::vector<TrtUniquePtrType<nvinfer1::IExecutionContext>>& exec_context);
+
+  /// Map input vector shapes to TRT Optimization profiles (min, max, opt)
+  // i.e. maps input_shapes_ to profiles_
+  void InitProfiles();
+
+  // Returns number of created profiles.
+  int GetNumProfiles() const;
+
+  // Restore profiles from the engine (used after deserialization)
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles.h]: Restores

File: tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles.h [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+
+  string DebugString() const {
+    using absl::StrCat;
+    return StrCat("[min: ", tensorflow::tensorrt::DebugString(min),
+                  ", opt: : ", tensorflow::tensorrt::DebugString(opt),
+                  ", max: ", tensorflow::tensorrt::DebugString(max), "]");
+  }
+
+#if IS_TRT_VERSION_GE(6, 0, 0, 0)
+  // Set the stored min/opt/max dimensions for profile.
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles.h]: Sets

File: tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles.h [hunk 13]
```diff
@@ -0,0 +1,178 @@
...
+      profile->setDimensions(name, nvinfer1::OptProfileSelector::kMAX, max[i]);
+    }
+    return Status::OK();
+  }
+#endif
+
+  // Returns true if profile range completely includes the given shapes.
+  bool IncludesShapes(const std::vector<TensorShape>& shapes) const {
+    // min, max, and opt must have the same size which,
+    // already verified in SetDimensions.
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles.h]: Avoid partial comment lines

File: tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles.h [hunk 14]
```diff
@@ -0,0 +1,178 @@
...
+    if (min.size() != shapes.size()) {
+      return false;
+    }
+    for (int i = 0; i < shapes.size(); i++) {
+      auto current_shape = shapes[i];
+      // min, max, and opt must have the same nbDims, which is
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles.h]: Avoid partial comment lines.

File: tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles.h [hunk 15]
```diff
@@ -0,0 +1,178 @@
...
+                          nvinfer1::IBuilderConfig* config,
+                          const nvinfer1::INetworkDefinition* network);
+#endif
+
+  // Creates execution contexts for each optimization profile.
+  Status CreateExecutionContexts(
+      nvinfer1::ICudaEngine* engine,
+      std::vector<TrtUniquePtrType<nvinfer1::IExecutionContext>>& exec_context);
+
+  /// Map input vector shapes to TRT Optimization profiles (min, max, opt)
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles.h]: Maps

File: tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles.h [hunk 16]
```diff
@@ -0,0 +1,178 @@
...
+ private:
+  // Set of input shape vetors that we collect during profile_generation_mode
+  std::unordered_set<std::vector<TensorShape>, VectorTensorShapeHasher>
+      input_shapes_;
+
+  // The optimization profiles generated from input_shapes_
+  std::vector<OptimizationProfileConfig> profiles_;
+
+#if IS_TRT_VERSION_GE(6, 0, 0, 0)
+  /// Add optimization profiles to the builder config
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles.h]: Adds

**Review by tfeher — COMMENTED** [review]
Thanks @bixia1 for the comments, I have addressed the issues.

**Review by bixia1 — APPROVED** [review]
Approving this so I have a convenient way to get it to my local tree to do some further checking.

File: tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles.h [hunk 17]
```diff
@@ -0,0 +1,178 @@
...
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#ifndef TENSORFLOW_COMPILER_TF2TENSORRT_CONVERT_SHAPE_OPTIMIZATION_PROFILES_H_
+#define TENSORFLOW_COMPILER_TF2TENSORRT_CONVERT_SHAPE_OPTIMIZATION_PROFILES_H_
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles.h]: TENSORFLOW_COMPILER_TF2TENSORRT_CONVERT_SHAPE_OPTIMIZATION_PROFILES_H_

The naming conversion here: path name + file name
TENSORFLOW_COMPILER_TF2TENSORRT_UTILS_TRT_SHAPE_OPTIMIZATION_PROFILES_H_

File: tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles.h [hunk 18]
```diff
@@ -0,0 +1,178 @@
...
+// input, and the optimal dimensions that the auto-tuner should use for
+// optimization.
+//
+// This class stores the list of input shapes that were seen during the
+// build/profile_generation_mode phase, and using them it creates a set of
+// OptimizationProfileConfigs. These configs will be added to IBuilderConfig
+// before the engine is created.
+class TrtShapeOptimizationProfile {
+ public:
+  TrtShapeOptimizationProfile(){};
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles.h]: Remove ;

File: tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles.cc [hunk 19]
```diff
@@ -0,0 +1,176 @@
...
+      if (i != idx) {
+        return errors::Internal(
+            "Profile index of engine config is different from resource profile "
+            "index: ",
+            i, " != ", idx);
+      }
+      VLOG(1) << "Added optimization profile " << profiles_[i].DebugString()
+              << " to builder config.";
+    } else {
+      VLOG(ERROR) << "Failed to add optimization profile "
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles.cc]: Does VLOG(ERROR) even pass compilation? It should be LOG(ERROR) if that is what you want.

File: tensorflow/python/compiler/tensorrt/test/trt_mode_test.py [hunk 20]
```diff
@@ -20,13 +20,13 @@
...
 from unittest import SkipTest  # pylint: disable=g-importing-member
 
+from tensorflow.compiler.tf2tensorrt.wrap_py_utils import get_linked_tensorrt_version
 from tensorflow.python.compiler.tensorrt.test import tf_trt_integration_test_base as trt_test
 from tensorflow.python.framework import dtypes
 from tensorflow.python.ops import array_ops
 from tensorflow.python.ops import math_ops
 from tensorflow.python.platform import test
 
-
```
Comment (bixia1) [inline tensorflow/python/compiler/tensorrt/test/trt_mode_test.py]: Why this line is deleted? It requires two lines.

File: tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles_test.cc [hunk 21]
```diff
@@ -0,0 +1,217 @@
...
+#include "third_party/tensorrt/NvInfer.h"
+
+namespace tensorflow {
+namespace tensorrt {
+
+std::vector<TensorShape> DimVecToShapeVec(std::vector<nvinfer1::Dims3> dimvec) {
+  std::vector<TensorShape> shapevec(dimvec.size());
+  for (int i = 0; i < dimvec.size(); i++) {
+    TensorShape shape;
+    TensorShapeUtils::MakeShape(dimvec[i].d, dimvec[i].nbDims, &shape);
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles_test.cc]: This returns a value, add TF_CHECK_OK(...)

File: tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles_test.cc [hunk 22]
```diff
@@ -0,0 +1,217 @@
...
+  TrtUniquePtrType<nvinfer1::IBuilderConfig> builder_config_;
+#endif
+  TrtUniquePtrType<nvinfer1::ICudaEngine> engine;
+  std::vector<TrtUniquePtrType<nvinfer1::IExecutionContext>> exec_context_;
+  // The order is important: exec_context_ must be destroyed first, and logger
+  // at last.
+
+  const uint32_t flags_ =
+      1U << static_cast<int>(
+          nvinfer1::NetworkDefinitionCreationFlag::kEXPLICIT_BATCH);
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles_test.cc]: This doesn't work for TRT5

File: tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles_test.cc [hunk 23]
```diff
@@ -0,0 +1,217 @@
...
+                           network_.get());
+
+  engine = TrtUniquePtrType<nvinfer1::ICudaEngine>(
+      builder_->buildEngineWithConfig(*network_, *builder_config_));
+#else
+  engine = TrtUniquePtrType<nvinfer1::ICudaEngine>(
+      builder_->buildCudaEngine(*network_));
+#endif
+  EXPECT_NE(nullptr, engine);
+  profile.CreateExecutionContexts(engine.get(), exec_context_);
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles_test.cc]: This returns a value, add TF_CHECK_OK(...)

File: tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles.cc [hunk 24]
```diff
@@ -0,0 +1,176 @@
...
+  // if TRT_VERSION < 6, then we do not need to add
+  return Status::OK();
+}
+#endif
+
+#if IS_TRT_VERSION_GE(6, 0, 0, 0)
+Status TrtShapeOptimizationProfile::ConfigureBuilder(
+    nvinfer1::IBuilder* builder, nvinfer1::IBuilderConfig* config,
+    const nvinfer1::INetworkDefinition* network) {
+  AddProfiles(builder, config, network);
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles.cc]: This returns a status, add  TF_RETURN_IF_ERROR(...)

File: tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles.cc [hunk 25]
```diff
@@ -0,0 +1,176 @@
...
+    i++;
+  } while (i < profiles_.size());
+
+  return Status::OK();
+}
+
+Status TrtShapeOptimizationProfile::RestoreProfiles(
+    const nvinfer1::ICudaEngine* engine) {
+#if IS_TRT_VERSION_GE(6, 0, 0, 0)
+  if (!engine || engine->hasImplicitBatchDimension()) {
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles.cc]: I see this compilation error no member named 'hasImplicitBatchDimension' in 'nvinfer1::ICudaEngine'

**Review by bixia1 — CHANGES_REQUESTED** [review]

**Review by tfeher — COMMENTED** [review]
Thanks @bixia1 for the comments! I have addressed the issues. I have confirmed that it works with TRT5, TRT6, and TRT7.

**Review by bixia1 — APPROVED** [review]
Approving it temporarily for myself to get it to my local tree.

File: tensorflow/compiler/tf2tensorrt/convert/utils.cc [hunk 26]
```diff
@@ -165,5 +165,25 @@ string GetLoadedTensorRTVersion() {
   return absl::StrCat(major, ".", minor, ".", patch);
 }
 
+int GetNumberOfEngineInputs(
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/convert/utils.cc]: Why this routine is outside #if GOOGLE_CUDA && GOOGLE_TENSORRT?
I am getting this compilation error
tensorflow/compiler/tf2tensorrt/convert/utils.cc:168:35: error: use of undeclared identifier 'nvinfer1'
int GetNumberOfEngineInputs(const nvinfer1::ICudaEngine* engine) {
                                  ^
tensorflow/compiler/tf2tensorrt/convert/utils.cc:179:5: error: function-like macro 'IS_TRT_VERSION_GE' is not defined
#if IS_TRT_VERSION_GE(6, 0, 0, 0)
    ^

when building target without --config=cuda tensorflow/compiler/tf2tensorrt:trt_shape_optimization_profiles_test

**Review by bixia1 — CHANGES_REQUESTED** [review]

File: tensorflow/compiler/tf2tensorrt/convert/utils.cc (hunk 26 above)
Comment (tfeher) [reply tensorflow/compiler/tf2tensorrt/convert/utils.cc]: That is a mistake, I have fixed it.

**Review by bixia1 — APPROVED** [review]

**Pull request merged**